File: android-reactive-location/src/main/java/pl/charmas/android/reactivelocation2/observables/ObservableEmitterWrapper.java
package pl.charmas.android.reactivelocation2.observables;

import io.reactivex.ObservableEmitter;
import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;

public class ObservableEmitterWrapper<T> implements Observer<T> {
    private final ObservableEmitter<T> emitter;

    public ObservableEmitterWrapper(ObservableEmitter<T> emitter) {
        this.emitter = emitter;
    }

    @Override
    public void onSubscribe(Disposable d) {
    }

    @Override
    public void onNext(T t) {
        if (!emitter.isDisposed()) {
            emitter.onNext(t);
        }
    }

    @Override
    public void onError(Throwable e) {
        if (!emitter.isDisposed()) {
            emitter.onError(e);
        }
    }

    @Override
    public void onComplete() {
        if (!emitter.isDisposed()) {
            emitter.onComplete();
        }
    }
}

File: android-reactive-location/src/main/java/pl/charmas/android/reactivelocation2/observables/ObservableEmitterWrapper.kt
package pl.charmas.android.reactivelocation2.observables

import io.reactivex.ObservableEmitter
import io.reactivex.Observer
import io.reactivex.disposables.Disposable

class ObservableEmitterWrapper<T>(private val emitter: ObservableEmitter<T>) :
    Observer<T> {
    override fun onSubscribe(d: Disposable) {}
    override fun onNext(value: T) {
        if (!emitter.isDisposed) {
            emitter.onNext(value)
        }
    }

    override fun onError(e: Throwable) {
        if (!emitter.isDisposed) {
            emitter.onError(e)
        }
    }

    override fun onComplete() {
        if (!emitter.isDisposed) {
            emitter.onComplete()
        }
    }
}